app/upgrades/v15/upgrades.go

```diff
@@ -59,6 +66,48 @@ func setICQParams(ctx sdk.Context, icqKeeper *icqkeeper.Keeper) {
 	icqKeeper.SetParams(ctx, icqparams)
 }
 
+func migrateBalancerPoolsToSolidlyStable(ctx sdk.Context, gammKeeper *gammkeeper.Keeper, poolmanagerKeeper *poolmanager.Keeper) {
+	// migrate stOSMO_OSMOPoolId, stJUNO_JUNOPoolId, stSTARS_STARSPoolId
+	pools := []uint64{stOSMO_OSMOPoolId, stJUNO_JUNOPoolId, stSTARS_STARSPoolId}
+	for _, poolId := range pools {
+		migrateBalancerPoolToSolidlyStable(ctx, gammKeeper, poolmanagerKeeper, poolId)
+	}
+}
+
+func migrateBalancerPoolToSolidlyStable(ctx sdk.Context, gammKeeper *gammkeeper.Keeper, poolmanagerKeeper *poolmanager.Keeper, poolId uint64) {
+	// fetch the pool with the given poolId
+	balancerPool, err := gammKeeper.GetPool(ctx, poolId)
+	if err != nil {
+		panic(err)
+	}
+
+	// initialize the stableswap pool
+	stableswapPool, err := stableswap.NewStableswapPool(
+		poolId,
+		stableswap.PoolParams{SwapFee: balancerPool.GetSwapFee(ctx), ExitFee: balancerPool.GetExitFee(ctx)},
+		balancerPool.GetTotalPoolLiquidity(ctx),
+		[]uint64{1, 1},
+		"osmo1k8c2m5cn322akk5wy8lpt87dd2f4yh9afcd7af", // Stride Foundation 2/3 multisig
+		"",
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	// ensure the number of stableswap LP shares is the same as the number of balancer LP shares
+	totalShares := sdk.NewCoin(
+		gammtypes.GetPoolShareDenom(poolId),
+		balancerPool.GetTotalShares(),
+	)
+	stableswapPool.TotalShares = totalShares
+
+	// overwrite the balancer pool with the new stableswap pool
+	err = gammKeeper.OverwritePool(ctx, &stableswapPool)
```

Comment:
I reckon we also need to handle the pool balance itself and migrate the pool balance to the new pool?

Reply:
Can you elaborate on this? Which field of which data structure needs to be migrated?

Reply:
Per https://github.com/osmosis-labs/osmosis/pull/4384/files#r1114879498, I think we should be good without this for now.

However, let's add an assert that the balances before and after are equal

Reply:
Is there a data structure that maps pool id -> balances or similar so that we don't have to iterate all addresses?

Reply:
We can use this to get an address https://github.com/osmosis-labs/osmosis/blob/a4e48c6d12620f64dc0c114985650e9544cd00c5/x/gamm/types/pool.go#L111-L115

Then, `bank.GetAllBalances(...)`

Reply:
Ah, I see! Thanks for clarifying, thought you meant checking LP share balances

Reply:
Done.